AI builder signup flow

apeatling opened the pull request:
This is the basic flow for Big Sky for free trial

1. Create account
2. Immediately launch a site
3. Redirect to Big Sky.

Still to do in this pr so Big SKy works

- [x] Set static home page
- [x] Set theme to assembler
- [ ] Properly do eligibility checks

### Testing instructions

1. Allow 3rd party cookies, for the site
2. Open fresh incognito session
3. Go to `http://calypso.localhost:3000/setup/ai-site-builder`
4. Go through Big Sky steps
5. Manually go to `http://calypso.localhost:3000/setup/ai-site-builder/domains?siteId=22760010`
6. Continue to checkout
7. Confirm that chosen plan and the domain is in cart.

Commit:
first pass

alshakero commented:
Hi Andy! I appreciate this PR because it reminded me of the acute need for fresh docs for our signup.

#### I'll address this step by step

##### The flow itself
[I made a sub-PR](https://github.com/Automattic/wp-calypso/pull/100953/) out of this and added verbose code comments in my this PR to explain my changes. 

#### The processing step

You don't need your own processing step, you can just `STEPS.PROCESSING`. If you want custom messages, you can add them [here](https://github.com/Automattic/wp-calypso/blob/aac1938d27cd7ddb534ba84095167f69daf514a3/client/landing/stepper/declarative-flow/internals/steps-repository/processing-step/hooks/use-processing-loading-messages.tsx#L14). 

I know you'll probably need to customize the created site. Ping me and we'l figure it out together. There are many store fields you can populate to customize the site.

daledupreez commented:
@apeatling, you may want to take a look at the example flow in `client/landing/stepper/declarative-flow/example.ts`. Have you had a chance to review that setup yet?

artpi commented:
I think one way would be to revert this PR, bring back AI-ASSEMBLER flow which was doing very similar things and then point it at big sky site.

alshakero commented:
> What can we improve in the existing APIs?

The foremost improvement that I have been putting off for the longest is the ability to pass props to steps directly from the flow. Something like

```ts
useStepsProps( step: stepSlug ) {
	switch ( step ) {
        case 'processing': 
			return {
	                phrases: [ __( 'making your site' ), ... ]
	         }
    }
}
```

The reason I delayed this is that I was unable to make solid types for this. But now that all the steps are [in the dictionary](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/steps.tsx), we can infer the step's types from its slug and thus 1) ensure safety 2) get free documentation of all the properties of the steps. And I'm working on inferring the flow's type from the return value of `initialize` (formerly useSteps) so if you pass props to a step that doesn't exist, types will also complain.

> Can we reduce the API surface so it's less possible to mess something up and easier to make it work?
1. We can remove all the deprecated code. Such use `useSteps`, `useStepNavigation.goBack`, `useStepNavigation.goToStep` and [DeprecatedStepperStep](https://github.com/Automattic/wp-calypso/blob/8401c290e07235c3f326eb94c880627e7345a37b/client/landing/stepper/declarative-flow/internals/types.ts#L45).
2. Once my effort [to move all the flows to Stepper auth ](https://github.com/Automattic/wp-calypso/pulls?q=sort%3Aupdated-desc+is%3Apr+is%3Aopen+label%3A%22Reviewer+Can+Merge%22+stepper)is done, remove `__experimentalUseBuiltinAuth` flag. I'm actively working on this now.
3. Once my effort to figure out the final shape of `useFlowState`, we can get rid of [`__experimentalUseSessions`](pdDR7T-242-p2). I'm actively working on this now.

> What additional tools can we borrow to make it more straightforward to build a new flow? Linting, types, tests, etc.?
Yesterday I learned a revelation (p1741200797324889-slack-CLCNELZE1) that will make typing Stepper flows much better.
1. You will only be able to `navigate` to steps that actually exist in your flow. Not just any string. That should improve robustness.
2. When you handle the submission of a given step, the value of `providedDependences` will be typed around that step. This will save us from many mistakes.
3. The flow will know all the steps it has, so we maybe able to infer all the state it needs. This is more of a vision IMO. I'm not working on this yet.

> What additional tools can we borrow to make it more straightforward to build a new flow? Linting, types, tests, etc.?
I believe types will help a LOT. We can also do quite a few runtime checks and throw clear errors when common mistakes are made. 

> What existing APIs or utils can we document or improve in terms of documentation?
This one is probably the most important.
1. I think we should have have 1-3 nice example flows that allow copy-pasting logic almost as-in, this will help AI too I suppose, we can leave code comments for AI to pick up.
2. I really want to make short videos 3-5 minutes explaining the general idea of Stepper. I think the learning curve is very steep but short. Once it clicks, it clicks.

> How can we make it easier for AI tooling to generate new flows and steps? Can we use the example flow as a baseline?
Frankly I don't know! I think leaving nice long and clear code comments may help.

alshakero committed:
Update ai-site-builder.ts (#100953)

* Update ai-site-builder.ts

* Delete client/landing/stepper/declarative-flow/internals/steps-repository/ai-site-builder-processing/index.tsx

You can add you custom messages in `client/landing/stepper/declarative-flow/internals/steps-repository/processing-step/hooks/use-processing-loading-messages.tsx`

* Discard changes to client/landing/stepper/declarative-flow/internals/steps.tsx

artpi committed:
Big Sky URL

artpi commented:
Companion on wpcom end: https://github.a8c.com/Automattic/wpcom/pull/176203

artpi committed:
Use big sky step instaed

apeatling commented:
@artpi Have you seen an issue with this flow and sandboxing public-api? I keep getting:

```
{
    "code": 400,
    "headers": [
        {
            "name": "Content-Type",
            "value": "application\/json"
        }
    ],
    "body": {
        "error": "invalid_key",
        "message": "Invalid activation key."
    }
}
```

Works fine without the API sandboxed. My sandbox is up to date.

kthai15 committed:
send continue site build email

kthai15 committed:
remove unneeded callback, add referrer to redirect

kthai15 committed:
add big-sky-free-trial blog sticker on launch

kthai15 commented:
630a279c97d93c2c6a6da5df05391e16dbac3970 adds the blog sticker on launch

will require this PR: https://github.a8c.com/Automattic/wpcom/pull/176836

apeatling committed:
Pass prompt to big sky

artpi committed:
Domains and plans steps

artpi commented:
@alshakero I was trying to figure it out today and I would appreciate help.
I did some of the work in https://github.com/Automattic/wp-calypso/pull/100948/commits/0fdd2f54d78c7a0bfb30261c86a51e276c7ed02a

What I am trying to accomplish is:
We go through `STEPS.SITE_CREATION_STEP`,`STEPS.PROCESSING`,`STEPS.LAUNCH_BIG_SKY` and then get redirected to Big Sky.

And then Big Sky would redirect to `STEPS.DOMAINS` with ?siteId=122&siteSlug=site.slug and then set the current site, and start searching a title.

What I am having trouble is:
- Current site gets already set in `STEPS.LAUNCH_BIG_SKY` so I understand it has to be set again back from the redirect? Should I create a new step to set it or do it in `STEPS.DOMAINS`  ?
- In `STEPS.DOMAINS` the `providedDependencies` are not at all passed when I use UNIFIED_DOMAINS so I switched to `STEPS.DOMAINS`. How would I store the chose domain if the site is already created?
- Similarly in plans. How would I persist the checkout choice before redirecting to checkout?

Currently the domain is not stored anywhere and the chosen plan as well.
I will continue digging tomorrow, but for now I am struggling with figuring out how the data actually gets passed around.

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,90 @@
+						true
+					);
+				}
+				case 'domains': {
+					// eslint-disable-next-line no-console
+					console.log( 'DOMAAAINZ STEP', providedDependencies );
+					// TODO: Somehow store the chosen domain.
+					return navigate( 'plans' );
+				}
+
+				case 'plans': {
+					// eslint-disable-next-line no-console
+					console.log( 'PLAAANZ STEP', providedDependencies );
+					// TODO: Somehow put the chosen plan in the cart.
+					const siteSlug = new URLSearchParams( window.location.search ).get( 'siteSlug' );
+					// eslint-disable-next-line no-console
+					console.log( 'SITE SLUG', siteSlug );
+					window.location.replace( `/checkout/${ encodeURIComponent( siteSlug || '' ) }` );
+				}
```
You don't have to store the domain. The domain step stores them automatically.  The domain step will add the product to the cart and to the store.

As for the plans, it only adds the plan to the store (not the cart), you can pick it up like [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/steps-repository/create-site/index.tsx#L78).

Then you can add it to the cart like [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/steps-repository/create-site/index.tsx).

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,90 @@
+						} );
+					} catch ( error ) {
+						// eslint-disable-next-line no-console
+						console.error( 'Failed to send continue build email:', error );
```
Here you can do `navigate ( 'error' )` We have a nice error step.

artpi replied:
We actually do not want to send these emails and this is for capturing the error and still continuing for the shape of this flow

alshakero commented:
> Current site gets already set in STEPS.LAUNCH_BIG_SKY so I understand it has to be set again back from the redirect? 
Yes, until the new state management is in prod. You can do it in an effect. 

```js
useSideEffect() {
	// Make sure to import import { useDispatch, useSelector } from 'calypso/state' (not the one from wordpress/data)
	const dispatch = useDispatch();
	const siteId = useQuery().get( 'siteId' );
	useEffect(() => {
		if( siteId ) {
			dispatch( setSelectedSiteId( siteId ) );
		}
	}, [ siteId ] )
}

> In STEPS.DOMAINS the providedDependencies are not at all passed when I use UNIFIED_DOMAINS so I switched to STEPS.DOMAINS. How would I store the chose domain if the site is already created?


> Similarly in plans. How would I persist the checkout choice before redirecting to checkout?

Addresses [here](https://github.com/Automattic/wp-calypso/pull/100948#discussion_r1993170785).

artpi committed:
Fixes from @alshakero recommendations

artpi committed:
cart item from provided dependencies

artpi committed:
Hide free plan for big sky

artpi committed:
Domain now works

artpi committed:
Flags for AI flow

kthai15 committed:
fix build issue

kthai15 committed:
remove add sticker

kthai15 committed:
cleanup

alshakero commented on client/landing/stepper/declarative-flow/registered-flows.ts:
```diff
@@ -99,6 +101,14 @@ const availableFlows: Record< string, () => Promise< { default: Flow } > > = {
 		import( /* webpackChunkName: "example-flow" */ '../declarative-flow/example' ),
 };
 
+const aiSiteBuilderFlows: Record< string, () => Promise< { default: Flow } > > = config.isEnabled(
+	'calypso/ai-site-builder-flow'
+)
+	? {
+			[ AI_SITE_BUILDER_FLOW ]: () => import( './ai-site-builder' ),
+	  }
+	: {};
+
```
I don't think you should bother with a flag, the flow has its own URL and isn't linked anywhere.

artpi replied:
Well, I wanna be safer just in case we are doing dumb stuff  in the flow, gonna remove it later

alshakero commented on client/landing/stepper/declarative-flow/registered-flows.ts:
```diff
@@ -17,6 +18,7 @@ import {
 	HUNDRED_YEAR_DOMAIN_FLOW,
 	EXAMPLE_FLOW,
 } from '@automattic/onboarding';
+import { AI_SITE_BUILDER_FLOW } from './ai-site-builder';
```
You have this const in the onboarding package.

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,147 @@
+import { Flow, ProvidedDependencies } from './internals/types';
+import type { OnboardSelect } from '@automattic/data-stores';
+
+export const AI_SITE_BUILDER_FLOW = 'ai-site-builder';
```
This const exists in the onboarding package.

artpi replied:
@kthai15 used it from there

kthai15 committed:
use AI_SITE_BUILDER_FLOW const in onboarding

artpi committed:
Merge branch 'trunk' into try/ai-builder-flow

artpi committed:
eligible only for the flow name

artpi commented on client/signup/steps/domains/index.jsx:
```diff
@@ -392,9 +392,17 @@ export class RenderDomainsStep extends Component {
 		const { step } = this.props;
-		const { suggestion } = step;
+
+		if ( step.suggestion ) {
```
@alshakero can I safely do this? For the existing flows, `step.suggestion` should work but it somehow isnt working for our flow so I had to pass it from line 271

ciampo replied:
> `step.suggestion` should work but it somehow isnt working for our flow

@alshakero may know better, but it looks like we should investigate (and fix) the expected behavior

alshakero replied:
Frankly I can't tell you if this is safe, what do you mean by not working, its not being submitted?

artpi replied:
Without this change it is `null`  and then it is null in providedDependencies too, which makes the domain not appear in checkout later.

alshakero replied:
I tried and honestly I can't reproduce. Anyways, don't be blocked by this but please give it another shot.

ciampo replied:
Now that this PR is merged, can we open a follow-up to clean this up?

artpi replied:
@ciampo I am not sure what do you mean by cleanup? I suspect this is the issue of the domain step not properly accounting for a situation when a domain step is launched AFTER website being created and the domain being added to the wrong cart (anonymous vs site specific). I honestly do not know how to properly fix it and account for other quirks of stepper.

artpi replied:
I'll try to debug this a bit more to maybe find more elegant solution https://github.com/Automattic/wp-calypso/issues/101516

artpi committed:
remove personal plan

artpi committed:
Typescript fixes

artpi committed:
stage flag

artpi commented:
@alshakero is it possible for you to have a look and tell me if we can merge it without worrying about other flows?
I cleaned it up a fair bit.
Could you review?

alshakero commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
 	}
 
+	getInitialQueryInAIFlow() {
+		if ( this.props.flowName !== 'ai-site-builder' ) {
```
Let's use the const from the onboarding package here. These consts help us a lot if we decide to remove the flow, as the const would lead us to all the conditions tied to it.

artpi replied:
https://github.com/Automattic/wp-calypso/pull/100948/commits/82ef7eb326213c02f38296ab314a690ba9e8b7cf

alshakero commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
 		return strippedHostname ?? hostname;
 	}
 
+	getInitialQueryInAIFlow() {
```
Maybe make this generic for better usability? Something like `getInitialQueryFromSiteName`. You can keep it only active for your flow but then people could add their flows as well.

artpi replied:
https://github.com/Automattic/wp-calypso/pull/100948/commits/82ef7eb326213c02f38296ab314a690ba9e8b7cf

alshakero commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
+			return;
+		}
+
+		// TODO: Maybe reading from meta or url would be useful too.
+		if ( ! this.props.selectedSite || ! this.props.selectedSite.name ) {
+			return;
+		}
+
+		return this.props.selectedSite.name;
```
```suggestion
		// TODO: Maybe reading from meta or url would be useful too.
		return this.props.selectedSite?.name;
```

ciampo replied:
[Jynx](https://github.com/Automattic/wp-calypso/pull/100948#discussion_r1995296660) !

In case returning `undefined` instead of `""` was important, we should also `|| undefined` at the end of the statement

artpi replied:
https://github.com/Automattic/wp-calypso/pull/100948/commits/82ef7eb326213c02f38296ab314a690ba9e8b7cf

alshakero commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
+			this.state.lastQuery ||
+			storedQuery ||
+			this.getInitialQueryInAIFlow() ||
+			this.getInitialQueryInLaunchFlow();
```
See, if this was generic, you would have just used it.

artpi replied:
Is that better?
https://github.com/Automattic/wp-calypso/pull/100948/commits/82ef7eb326213c02f38296ab314a690ba9e8b7cf

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,153 @@
+					}
+
+					// get the prompt from the get url
+					const prompt = new URLSearchParams( window.location.search ).get( 'prompt' );
```
You have `useQuery` imported. Might as well?

artpi replied:
Done https://github.com/Automattic/wp-calypso/pull/100948/commits/678834230db0a7215557183c0c3bae970f25915f

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,153 @@
+					let promptParam = '';
+
+					if ( prompt ) {
+						promptParam = `&prompt=${ prompt }`;
```
Should probably be escaped because this is user input
```suggestion
						promptParam = `&prompt=${ encodeURIComponent( prompt ) }`;
```

artpi replied:
Done https://github.com/Automattic/wp-calypso/pull/100948/commits/678834230db0a7215557183c0c3bae970f25915f

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,153 @@
+					console.log( 'DOMAAAINZ STEP', providedDependencies, siteSlugFromSiteData );
+					// TODO: Somehow store the chosen domain.
+					if ( providedDependencies.domainItem && siteSlugFromSiteData ) {
+						const addToCart = await addProductsToCart( siteSlugFromSiteData, AI_SITE_BUILDER_FLOW, [
```
IMHO, you don't have to await this, let it run in the background while the user browses for a plan. Because now they will pick a domain, and it will freeze for a second before showing the plans step for no good reason.

artpi replied:
Done!

alshakero commented on client/landing/stepper/declarative-flow/ai-site-builder.ts:
```diff
@@ -0,0 +1,153 @@
+					}
+
+					// eslint-disable-next-line no-console
+					window.location.replace(
```
You can use `window.location.assign`, otherwise clicking back from checkout will take the user to domains step.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/launch-big-sky/index.tsx:
```diff
@@ -104,7 +108,9 @@ const LaunchBigSky: Step = function () {
 				window.location.replace(
-					`${ siteURL }/wp-admin/site-editor.php?canvas=edit&referrer=design-choices`
+					`${ siteURL }/wp-admin/site-editor.php?canvas=edit&referrer=${
+						referrer || 'design-choices'
```
This should be escaped as well, because it's user input. Probably even better if we only allow a list of options.

artpi replied:
I simplified it much more - we passed the flow name in the launch big sky step.
https://github.com/Automattic/wp-calypso/pull/100948/commits/678834230db0a7215557183c0c3bae970f25915f

ciampo commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
 	}
 
+	getInitialQueryInAIFlow() {
+		if ( this.props.flowName !== 'ai-site-builder' ) {
+			return;
+		}
```
Checking the flow name makes a step less reusable and more coupled with flow logic — could we do without this check?

And otherwise, can we reuse the `AI_SITE_BUILDER_FLOW` constant instead of the hardcoded string?

ciampo commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -337,9 +337,26 @@ class RegisterDomainStep extends Component {
+		}
+
+		// TODO: Maybe reading from meta or url would be useful too.
+		if ( ! this.props.selectedSite || ! this.props.selectedSite.name ) {
+			return;
+		}
+
+		return this.props.selectedSite.name;
```
```suggestion
		return this.props.selectedSite?.name || undefined;
```

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/index.tsx:
```diff
@@ -127,10 +128,19 @@ export default function PlansStepAdaptor( props: StepProps ) {
-			: getHidePlanPropsBasedOnThemeType( selectedThemeType || '' );
+	let hidePlanProps;
+	if ( createWithBigSky && isGoalFirstExperiment ) {
+		hidePlanProps = getHidePlanPropsBasedOnCreateWithBigSky();
```
Can we extend this function to do it instead? It seems related.

artpi replied:
Honestly I struggle to make it reusable without adding more complexity than it already has. The function would need 3 parameters to return 3 lines and I don't know if it's worth it.
Here is how it shakes out:

```
	let hidePlanProps;
	if ( createWithBigSky && isGoalFirstExperiment ) {
		hidePlanProps = {
			hideFreePlan: true,
			hideEcommercePlan: true,
			hideEnterprisePlan: true,
		};
	} else if ( props.flow === AI_SITE_BUILDER_FLOW ) {
		hidePlanProps = {
			hideFreePlan: true,
			hidePersonalPlan: true,
			hideEcommercePlan: true,
			hideEnterprisePlan: true,
		};
	} else {
		hidePlanProps = getHidePlanPropsBasedOnThemeType( selectedThemeType || '' );
	}
	```

alshakero reviewed (commented):
Besides [this](https://github.com/Automattic/wp-calypso/pull/100948/files#r1994398934) (which I don't think is bad, just not sure about it), you can merge this safely. 

I did leave some feedback, while non-blocking, it would be great if you address it because these parts of the code are really prone to deterioration.

artpi committed:
Update client/components/domains/register-domain-step/index.jsx

Co-authored-by: Marco Ciampini <marco.ciampini@automattic.com>

artpi committed:
More generic query fallbacks from PR feedback

artpi committed:
Simplify referer passing

artpi committed:
remove await

artpi committed:
window.location.assign

alshakero commented:
I checked and it's safe to merge. The domains step works in other flows.

kthai15 reviewed (approved):
works for me, tested this flow and the existing design-choices flow
